File: FMELibrary/BinaryParser.cs
using System.Drawing;
using System.Text;

namespace FMELibrary
{
    /// <summary>
    /// Provides extension methods for reading and writing binary data with custom serialization logic.
    /// </summary>
    public static class BinaryParser
    {
        /// <summary>
        /// Reads a length-prefixed string from the binary reader.
        /// The first 4 bytes represent the string length.
        /// </summary>
        /// <param name="reader">The binary reader to read from.</param>
        /// <returns>A UTF-8 decoded string.</returns>
        public static string ReadStringEx(this BinaryReader reader)
        {
            int length = reader.ReadInt32();
            var bytes = reader.ReadBytes(length);
            return Encoding.UTF8.GetString(bytes);
        }

        /// <summary>
        /// Reads a 16-bit RGB565 color value and converts it to a <see cref="Color"/> object.
        /// </summary>
        /// <param name="reader">The binary reader to read from.</param>
        /// <returns>A <see cref="Color"/> object representing the RGB color.</returns>
        public static Color ReadColor(this BinaryReader reader)
        {
            short color = reader.ReadInt16();
            int red = color >> 10;
            int green = (color >> 5) & 0x1F;
            int blue = color & 0x1F;

            return Color.FromArgb(red << 3, green << 3, blue << 3);
        }

        public static DateOnly ReadDate(this BinaryReader reader)
        {
            int dayOfYear = reader.ReadInt16();
            int year = reader.ReadInt16();
            if (dayOfYear == -1 || year == -1)
                return DateOnly.MinValue;

            return new DateOnly(year, 1, 1).AddDays(dayOfYear - 1);
        }

        /// <summary>
        /// Writes various data types to the binary writer with automatic type detection and appropriate serialization.
        /// </summary>
        /// <param name="writer">The binary writer to write to.</param>
        /// <param name="value">The value to write. Supported types: byte, short, int, float, byte[], string, Color, and null.</param>
        /// <exception cref="InvalidOperationException">Thrown when the value type is not supported.</exception>
        public static void WriteEx(this BinaryWriter writer, object? value)
        {
            switch (value)
            {
                case bool b:
                    writer.Write(b);
                    break;

                case byte b:
                    writer.Write(b);
                    break;

                case short s:
                    writer.Write(s);
                    break;

                case int i:
                    writer.Write(i);
                    break;

                case long l:
                    writer.Write(l);
                    break;

                case float f:
                    writer.Write(f);
                    break;

                case byte[] b:
                    writer.Write(b);
                    break;

                case string s:
                    var bytes = Encoding.UTF8.GetBytes(s);
                    writer.Write(bytes.Length);
                    writer.Write(bytes);
                    break;

                case Color c:
                    int red = c.R >> 3;
                    int green = c.G >> 3;
                    int blue = c.B >> 3;
                    writer.Write((short)((red << 10) | (green << 5) | blue));
                    break;

                case DateOnly d:
                    int dayOfYear = d == DateOnly.MinValue ? -1 : d.DayOfYear;
                    int year = d == DateOnly.MinValue ? -1 : d.Year;
                    writer.Write((short)dayOfYear);
                    writer.Write((short)year);
                    break;

                case null:
                    break;

                default:
                    throw new InvalidOperationException($"Unsupported type: {value?.GetType().FullName}");
            }
        }
    }
}

File: FMELibrary.js/BinaryParser.ts
/**
 * Provides utility functions for reading and writing binary data with custom serialization logic.
 */
export class BinaryParser {
  /**
* Reads a UTF-8 encoded string of the specified length from the buffer.
   * @param buffer - The buffer to read from
   * @param offset - The current offset in the buffer
   * @param length - The number of bytes to read
   * @returns An object containing the decoded string and the new offset
 */
  static readString(buffer: Buffer, offset: number, length: number): { value: string; offset: number } {
    const value = buffer.toString('utf-8', offset, offset + length);
    return { value, offset: offset + length };
  }

  /**
   * Reads a length-prefixed string from the buffer. The first 4 bytes represent the string length.
   * @param buffer - The buffer to read from
   * @param offset - The current offset in the buffer
   * @returns An object containing the decoded string and the new offset
   */
  static readStringEx(buffer: Buffer, offset: number): { value: string; offset: number } {
    const length = buffer.readInt32LE(offset);
    offset += 4;
    return this.readString(buffer, offset, length);
  }

  /**
   * Reads a 16-bit RGB565 color value and converts it to RGB components.
   * @param buffer - The buffer to read from
   * @param offset - The current offset in the buffer
* @returns An object containing the RGB color values and the new offset
   */
  static readColor(buffer: Buffer, offset: number): { r: number; g: number; b: number; offset: number } {
    const color = buffer.readInt16LE(offset);
    const red = color >> 10;
    const green = (color >> 5) & 0x1F;
    const blue = color & 0x1F;

    return {
      r: red << 3,
      g: green << 3,
      b: blue << 3,
      offset: offset + 2
    };
  }

  /**
   * Writes various data types to a buffer with automatic type detection and appropriate serialization.
* @param value - The value to write. Supported types: number (byte, short, int, float), Buffer, string, color object, null/undefined
   * @param buffer - The buffer to write to
   * @param offset - The current offset in the buffer
   * @returns The new offset after writing
   * @throws Error when the value type is not supported
   */
  static writeEx(value: any, buffer: Buffer, offset: number): number {
    if (value === null || value === undefined) {
      return offset;
    }

    if (typeof value === 'number') {
      // Determine size based on context - caller should specify type
      buffer.writeInt32LE(value, offset);
      return offset + 4;
    }

    if (Buffer.isBuffer(value)) {
      value.copy(buffer, offset);
      return offset + value.length;
    }

    if (typeof value === 'string') {
      const bytes = Buffer.from(value, 'utf-8');
      buffer.writeInt32LE(bytes.length, offset);
      offset += 4;
      bytes.copy(buffer, offset);
      return offset + bytes.length;
    }

    if (typeof value === 'object' && 'r' in value && 'g' in value && 'b' in value) {
      const red = value.r >> 3;
      const green = value.g >> 3;
      const blue = value.b >> 3;
      buffer.writeInt16LE((red << 10) | (green << 5) | blue, offset);
      return offset + 2;
    }

    throw new Error(`Unsupported type: ${typeof value}`);
  }

  /**
   * Writes a byte value to the buffer.
   */
  static writeByte(value: number, buffer: Buffer, offset: number): number {
    buffer.writeUInt8(value, offset);
    return offset + 1;
  }

  /**
     * Writes a short (16-bit integer) value to the buffer.
     */
  static writeInt16(value: number, buffer: Buffer, offset: number): number {
    buffer.writeInt16LE(value, offset);
    return offset + 2;
  }

  /**
   * Writes an int (32-bit integer) value to the buffer.
   */
  static writeInt32(value: number, buffer: Buffer, offset: number): number {
    buffer.writeInt32LE(value, offset);
    return offset + 4;
  }

  /**
   * Writes a float (32-bit floating point) value to the buffer.
   */
  static writeFloat(value: number, buffer: Buffer, offset: number): number {
    buffer.writeFloatLE(value, offset);
    return offset + 4;
  }
}
